### PR TITLE
add image analysis w/ tensorflow

### PR DESCRIPTION
**Pull request by h324yang**
# JCDL2019 demo
Using AUT and SSD model w/ Tensorflow to do object detection analysis on web archives.
* * *
1. default setting is `standalone` mode, so need to set up master and slaves first.
2. run `detect.py` to get and store the object probabilities and the image byte strings.
3. run `extract_images.py` to get image files from the result of `step2`

### CODE REVIEW

**Comment by h324yang** [comment]
# Distributed image analysis via the integration of AUT and Tensorflow

**GitHub issue(s)**: #240 #241

# What does this Pull Request do?

* Integrating AUT and Tensorflow with python interface (pyspark). 
* The code of the JCDL 2019 paper.
* Single Shot MultiBox Detector is used so far, because of the balance between speed and accuracy.
* The inference scores and the byte strings of images are stored first.
* Using the image extractor to get the image files, , i.e., jpeg, gif, etc., which scores are higher than the threshold defined by users.

# How should this be tested?
#### Step 1: Run detection
	python aut/src/main/python/tf/detect.py \
			--web_archive "/tuna1/scratch/nruest/geocites/warcs/1/*" \
			--aut_jar aut/target/aut-0.17.1-SNAPSHOT-fatjar.jar \
			--aut_py aut/src/main/python \
			--spark spark-2.3.2-bin-hadoop2.7/bin \
			--master spark://127.0.1.1:7077 \
			--img_model ssd \
			--filter_size 640 640 \
			--output_path warc_res

#### Step 2: Extract Images
	python aut/src/main/python/tf/extract_images.py \
			--res_dir warc_res \
			--output_dir warc_imgs \
			--threshold 0.85

# Additional Notes:

#### Python Dependency
My python environment is as listed in [here](https://github.com/h324yang/warc-image-classification/blob/master/req.txt). Though it's not the minimal requirement, to quickly set up, you can directly download it and then  `pip install req.txt` .

Note that you should ensure that  driver and workers use the same python version. You might set as follows:

	export PYSPARK_PYTHON=[YOUR PYTHON]
	export PYSPARK_DRIVER_PYTHON=[YOUR PYTHON]

#### Spark Mode
The default mode is `standalone`. E.g., you can launch in this mode as follows:

	cd spark-2.3.2-bin-hadoop2.7
	./sbin/start-master.sh
	./start-slave.sh 127.0.1.1:7077

The spark parameters are set by using `init_spark()` in `src/main/python/tf/util/init.py`

#### Design Details
- The pre-trained model and the corresponding dictionary for label mapping are stored in `src/main/python/tf/model/graph/` and `src/main/python/tf/model/category/` , respectively.
- For each pre-trained model, though there is only one now, we define a model class and an extractor class, as `SSD` and `SSDExtractor` in `src/main/python/tf/model/object_detection.py`. 
- Using the model class, as `SSD`, to derive the pandas UDF function for inference. 

# Interested parties

@lintool

File: src/main/python/tf/util/init.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    sql_context = SQLContext(sc)
+    return conf, sc, sql_context
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description='PySpark for Web Archive Image Retrieval')
+    parser.add_argument('--web_archive', help='input directory for web archive data', default='/tuna1/scratch/nruest/geocites/warcs')
+    parser.add_argument('--aut_jar', help='aut compiled jar package', default='aut/target/aut-0.17.1-SNAPSHOT-fatjar.jar')
+    parser.add_argument('--aut_py', help='path to python package', default='aut/src/main/python')
+    parser.add_argument('--spark', help='path to python package', default='spark-2.3.2-bin-hadoop2.7/bin')
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: Change to: `help='Path to Apache Spark.`

File: src/main/python/tf/util/init.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    sc = SparkContext(master, "aut image analysis", conf=conf)
+    sql_context = SQLContext(sc)
+    return conf, sc, sql_context
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description='PySpark for Web Archive Image Retrieval')
+    parser.add_argument('--web_archive', help='input directory for web archive data', default='/tuna1/scratch/nruest/geocites/warcs')
+    parser.add_argument('--aut_jar', help='aut compiled jar package', default='aut/target/aut-0.17.1-SNAPSHOT-fatjar.jar')
+    parser.add_argument('--aut_py', help='path to python package', default='aut/src/main/python')
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: Is this supposed to be the Python binary? Or something else? Needs a better help description.

File: src/main/python/tf/util/init.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    conf.set("spark.task.cpus", "3")
+    sc = SparkContext(master, "aut image analysis", conf=conf)
+    sql_context = SQLContext(sc)
+    return conf, sc, sql_context
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description='PySpark for Web Archive Image Retrieval')
+    parser.add_argument('--web_archive', help='input directory for web archive data', default='/tuna1/scratch/nruest/geocites/warcs')
+    parser.add_argument('--aut_jar', help='aut compiled jar package', default='aut/target/aut-0.17.1-SNAPSHOT-fatjar.jar')
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: Change to: `help='Path to compiled aut jar.'

File: src/main/python/tf/util/init.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    conf.set("spark.driver.memory", "64G")
+    conf.set("spark.task.cpus", "3")
+    sc = SparkContext(master, "aut image analysis", conf=conf)
+    sql_context = SQLContext(sc)
+    return conf, sc, sql_context
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description='PySpark for Web Archive Image Retrieval')
+    parser.add_argument('--web_archive', help='input directory for web archive data', default='/tuna1/scratch/nruest/geocites/warcs')
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: Change to: `help='Path to warcs.'`

File: src/main/python/tf/util/init.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    parser.add_argument('--aut_jar', help='aut compiled jar package', default='aut/target/aut-0.17.1-SNAPSHOT-fatjar.jar')
+    parser.add_argument('--aut_py', help='path to python package', default='aut/src/main/python')
+    parser.add_argument('--spark', help='path to python package', default='spark-2.3.2-bin-hadoop2.7/bin')
+    parser.add_argument('--master', help='master IP address', default='spark://127.0.1.1:7077')
+    parser.add_argument('--img_model', help='model for image processing, use ssd', default='ssd')
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: > model for image processing, use ssd

If this is the only option, why is there an argument?

File: src/main/python/tf/util/init.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    parser.add_argument('--filter_size', nargs='+', type=int, help='filter out images smaller than filter_size', default=[640, 640])
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: Change to: `help='Filter out images smaller than filter_size.'`

File: src/main/python/tf/util/init.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    parser.add_argument('--output_path', help='image model output dir', default='coco_res')
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: Change to: `help='Path to image model output.'`

File: src/main/python/tf/util/init.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    return conf, sc, sql_context
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description='PySpark for Web Archive Image Retrieval')
+    parser.add_argument('--web_archive', help='input directory for web archive data', default='/tuna1/scratch/nruest/geocites/warcs')
+    parser.add_argument('--aut_jar', help='aut compiled jar package', default='aut/target/aut-0.17.1-SNAPSHOT-fatjar.jar')
+    parser.add_argument('--aut_py', help='path to python package', default='aut/src/main/python')
+    parser.add_argument('--spark', help='path to python package', default='spark-2.3.2-bin-hadoop2.7/bin')
+    parser.add_argument('--master', help='master IP address', default='spark://127.0.1.1:7077')
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: Change to: `help='Apache Spark master IP address and port.'`

File: src/main/python/tf/util/init.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+import argparse
+import os
+import zipfile
+from pyspark import SparkConf, SparkContext, SQLContext
+
+
+def init_spark(master, aut_jar):
```
Comment (ruebot) [inline src/main/python/tf/util/init.py]: Should these all be hard coded Spark confs?

File: src/main/python/tf/extract_images.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    parser.add_argument('--threshold', type=float, help='threshold of detection confidence scores')
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    args = get_args()
+    extractor = SSDExtractor(args.res_dir, args.output_dir)
+    extractor.extract_and_save(class_ids="all", threshold=args.threshold)
+
+
```
Comment (ruebot) [inline src/main/python/tf/extract_images.py]: Remove unnecessary end lines.

File: src/main/python/tf/extract_images.py [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import numpy as np
+import argparse
+from model.object_detection import SSDExtractor
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description='Extracting images from model output')
```
Comment (ruebot) [inline src/main/python/tf/extract_images.py]: Change to: `description='Extracting images from model output.'`

File: src/main/python/tf/extract_images.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    parser.add_argument('--res_dir', help='result (model output) dir')
```
Comment (ruebot) [inline src/main/python/tf/extract_images.py]: Change to: `help='Path of result (model output) directory.'`

File: src/main/python/tf/extract_images.py [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+    parser.add_argument('--output_dir', help='extracted image file output dir')
```
Comment (ruebot) [inline src/main/python/tf/extract_images.py]: Change to: `help='Path of extracted image file output directory.'`

File: src/main/python/tf/extract_images.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+    parser.add_argument('--threshold', type=float, help='threshold of detection confidence scores')
```
Comment (ruebot) [inline src/main/python/tf/extract_images.py]: Change to: `help='Threshold of detection confidence scores.'`

**Comment by ianmilligan1** [comment]
All of the options sound good to me for various reasons! But I think at this stage as a prototype function we could probably just have people add some flags and roll with it – down the line, perhaps as a separate issue, come up with a `conf` file to try to reduce some of the flag soup? @ruebot

**Comment by h324yang** [comment]
> @h324yang did y'all get a lot of [this](https://gist.github.com/ruebot/b0d59e0f1bb5bb342f34ba377b88560c) when you ran the first pass script? Just trying to understand what's normal/expected behaviour here.

Seems like an OOM error; The arguments I set in util/init.py were optimized and running well on Tuna. I got some errors but I don't think OOM is a frequent one. You also run on Tuna? 

Maybe a lower value of "spark.sql.execution.arrow.maxRecordsPerBatch" could help, e.g., 1280 -> 640. (Indeed, tuning such settings bothered me a lot :-/)

File: src/main/python/tf/util/init.py (hunk 6 above)
Comment (ruebot) [reply src/main/python/tf/util/init.py]: `default=[640, 640]` should be `default=[640 640]`, right? That comma will cause an error.

File: src/main/python/tf/util/init.py (hunk 5 above)
Comment (h324yang) [reply src/main/python/tf/util/init.py]: I expect that we may have other models in the future, though we only have this one now.

File: src/main/python/tf/util/init.py (hunk 6 above)
Comment (h324yang) [reply src/main/python/tf/util/init.py]: I think it works fine. The comma in python code just makes the default argument a list of int. While using command line Interface, neither the comma nor the brackets are needed. We just use space as the delimiter, e.g. `python detect.py --filter_size 100 100`.

File: src/main/python/tf/util/init.py (hunk 2 above)
Comment (h324yang) [reply src/main/python/tf/util/init.py]: This is the path to AUT python package, in order to import `aut.common`. I found this is not necessary though. It's already hard coded in `detect.py` (line 5,6). So I think I just delete this to avoid confusion.

**Comment by h324yang** [comment]
> We might want to address this message from when we run the initial pass too:
> 
> ```
> WARNING:tensorflow:From /home/ruestn/aut/src/main/python/tf/model/object_detection.py:49: FastGFile.__init__ (from tensorflow.python.platform.gfile) is deprecated and will be removed in a future version.
> ```

I update to the TF 1.14.0 api, i.e. `tf.io.gfile.GFile`.

File: src/main/python/tf/util/init.py (hunk 9 above)
Comment (h324yang) [reply src/main/python/tf/util/init.py]: I add a `conf` file, i.e. `util/spark.conf`, so users can list everything they'd like to set up there. Hopefully that is better?

**Comment by h324yang** [comment]
@ruebot I done all requested changes except for `--img_model`, which reason is replied in the thread. Also, `conf` file is added. Please re-review the new commits.

**Comment by h324yang** [comment]
Sorry! That slipped my mind, and I already removed it.
The model is from [TF detection model zoo](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md): ssd_mobilenet_v1_fpn_coco ☆

We can download it and mv the `frozen_inference_graph.pb` to the designated folder `aut/src/main/python/tf/model/graph/ssd_mobilenet_v1_fpn_640x640`

For example:
```
wget http://download.tensorflow.org/models/object_detection/ssd_mobilenet_v1_fpn_shared_box_predictor_640x640_coco14_sync_2018_07_03.tar.gz
tar -xzvf ssd_mobilenet_v1_fpn_shared_box_predictor_640x640_coco14_sync_2018_07_03.tar.gz
mkdir -p aut/src/main/python/tf/model/graph/ssd_mobilenet_v1_fpn_640x640/
cp ssd_mobilenet_v1_fpn_shared_box_predictor_640x640_coco14_sync_2018_07_03/frozen_inference_graph.pb aut/src/main/python/tf/model/graph/ssd_mobilenet_v1_fpn_640x640/
```
